File: thesis/algorithms/python/np.py
import sys

def parse_numbers(path):
    with open(path, "r") as f:
        content = f.read().strip()
    content = content.replace('[', '').replace(']', '')
    return [int(x.strip()) for x in content.split(',') if x.strip()]

def find_factors(n):
    for i in range(2, int(n ** 0.5) + 1):
        if n % i == 0:
            return (i, n // i)
    return None

def main():
    if len(sys.argv) < 2:
        print("Uso: python3 np.py <tamanho>")
        return

    tamanho = sys.argv[1]
    path = f"datasets/{tamanho}/factoring.json"

    try:
        numbers = parse_numbers(path)

        total = len(numbers)
        fatorados = 0
        primos = 0

        for n in numbers:
            result = find_factors(n)
            if result:
                fatorados += 1
            else:
                primos += 1

        print(f"Fatorados: {fatorados}")


    except Exception as e:
        print(f"Erro: {e}")

if __name__ == "__main__":
    main()

File: thesis/algorithms/typescript/np.ts
import * as fs from 'fs';
import * as path from 'path';

function parseNumbers(filePath: string): number[] {
  const content = fs.readFileSync(filePath, 'utf-8').trim();
  const cleaned = content.replace('[', '').replace(']', '');
  return cleaned
    .split(',')
    .map((s) => parseInt(s.trim()))
    .filter((n) => !isNaN(n));
}

function findFactors(n: number): [number, number] | null {
  for (let i = 2; i <= Math.floor(Math.sqrt(n)); i++) {
    if (n % i === 0) {
      return [i, n / i];
    }
  }
  return null;
}

function main() {
  const args = process.argv.slice(2);
  if (args.length < 1) {
    console.log('Uso: tsx np.ts <tamanho>');
    return;
  }

  const tamanho = args[0];
  const filePath = path.join('datasets', tamanho, 'factoring.json');

  if (!fs.existsSync(filePath)) {
    console.error(`Arquivo não encontrado: ${filePath}`);
    return;
  }

  try {
    const numbers = parseNumbers(filePath);
    let fatorados = 0;
    let primos = 0;

    for (const n of numbers) {
      if (findFactors(n)) {
        fatorados++;
      } else {
        primos++;
      }
    }

    console.log(`Fatorados: ${fatorados}`);
  } catch (err) {
    console.error(`Erro: ${(err as Error).message}`);
  }
}

main();
